File: scripts/fix-specs.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Use sync glob since async might not be available
function findFiles(dir, pattern) {
  const result = [];
  const files = fs.readdirSync(dir);

  for (const file of files) {
    const fullPath = path.join(dir, file);
    const stat = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      result.push(...findFiles(fullPath, pattern));
    } else if (file.match(pattern)) {
      result.push(fullPath);
    }
  }

  return result;
}

const SPEC_DIR = path.join(__dirname, '../tests/specFiles/ga');

const stats = {
  total: 0,
  fixed: 0,
  skipped: 0,
  errors: 0,
};

function getImportDepth(filePath) {
  const parts = filePath.split(path.sep);
  const gaIndex = parts.indexOf('ga');
  if (gaIndex >= 0) {
    // Count how many segments after ga
    const afterGa = parts.length - gaIndex - 1;
    // ga/component/file.spec.ts = 2 levels = ../../../
    // ga/file.spec.ts = 1 level = ../../
    return afterGa > 1 ? '../../../' : '../../';
  }
  return '../../../';
}

function fixSpecFile(filePath) {
  try {
    let content = fs.readFileSync(filePath, 'utf-8');
    const original = content;

    // 1. Add report-enhancer import if missing
    if (!content.includes('attachConsoleCapture')) {
      const depth = getImportDepth(filePath);
      const importLine = `import { attachConsoleCapture, annotateEnvironment } from '${depth}utils/infra/report-enhancer';`;

      // Find last import
      const importMatches = [];
      const importRegex = /^import .* from ['""].+['"""];?$/gm;
      let match;
      while ((match = importRegex.exec(content)) !== null) {
        importMatches.push(match);
      }

      if (importMatches.length > 0) {
        const lastImport = importMatches[importMatches.length - 1];
        const insertPos = lastImport.index + lastImport[0].length;
        if (!content.includes(importLine)) {
          content = content.slice(0, insertPos) + '\n' + importLine + content.slice(insertPos);
        }
      }
    }

    // 2. Add capture variable if missing
    if (!content.includes('let capture: ConsoleCapture;')) {
      // Find insertion point
      const match = content.match(/\n(const [A-Z_]+ = |test\.)/);
      if (match && match.index) {
        const insertPos = match.index + 1;
        content = content.slice(0, insertPos) + 'let capture: ConsoleCapture;\n\n' + content.slice(insertPos);
      }
    }

    // 3. Update beforeEach
    const beforeEachRegex = /test\.beforeEach\s*\(\s*async\s*\(\s*{\s*page\s*}\s*\)\s*=>\s*{/;
    if (beforeEachRegex.test(content) && !content.includes('capture = new ConsoleCapture')) {
      // Find the closing brace of beforeEach
      const beforeEachMatch = content.match(beforeEachRegex);
      if (beforeEachMatch) {
        let braceCount = 1;
        let pos = beforeEachMatch.index + beforeEachMatch[0].length;
        while (braceCount > 0 && pos < content.length) {
          if (content[pos] === '{') braceCount++;
          if (content[pos] === '}') braceCount--;
          pos++;
        }
        // Insert before the closing });
        const insertPos = pos - 1;
        const captureInit = '\n  capture = new ConsoleCapture(page);\n  capture.start();';
        content = content.slice(0, insertPos) + captureInit + content.slice(insertPos);
      }
    }

    // 4. Add afterEach if missing
    if (!content.includes('test.afterEach')) {
      // Find where to add (after beforeEach)
      const beforeEachStart = content.indexOf('test.beforeEach');
      const closePos = content.indexOf('});', beforeEachStart);
      if (closePos > 0) {
        const insertPos = closePos + 3;
        const afterEachBlock = `

test.afterEach(async ({ page }, testInfo) => {
  const errors = capture.getErrors();
  const warnings = capture.getWarnings();
  if (errors.length > 0 || warnings.length > 0) {
    await attachConsoleCapture(page, testInfo, errors, warnings);
  }
  await annotateEnvironment(page, testInfo);
});`;
        content = content.slice(0, insertPos) + afterEachBlock + content.slice(insertPos);
      }
    }

    if (content !== original) {
      fs.writeFileSync(filePath, content, 'utf-8');
      stats.fixed++;
      const rel = path.relative(process.cwd(), filePath);
      console.log(`✓ Fixed: ${rel}`);
    } else {
      stats.skipped++;
    }
  } catch (error) {
    stats.errors++;
    console.error(`✗ Error: ${path.relative(process.cwd(), filePath)}: ${error.message}`);
  }
}

function main() {
  console.log('🔧 Framework Refactoring Script\n');
  console.log(`Scanning: ${SPEC_DIR}\n`);

  // Find all .author.spec.ts files
  const authorSpecs = findFiles(SPEC_DIR, /\.author\.spec\.ts$/);

  console.log(`Found ${authorSpecs.length} author specs\n`);

  authorSpecs.forEach(filePath => {
    stats.total++;
    fixSpecFile(filePath);
  });

  // Print summary
  console.log(`\n${'='.repeat(60)}`);
  console.log('📊 Summary:');
  console.log(`   Total scanned:  ${stats.total}`);
  console.log(`   Fixed:          ${stats.fixed}`);
  console.log(`   Skipped:        ${stats.skipped}`);
  console.log(`   Errors:         ${stats.errors}`);
  console.log(`${'='.repeat(60)}`);

  process.exit(stats.errors > 0 ? 1 : 0);
}

main();

File: scripts/fix_specs.py
#!/usr/bin/env python3

"""
Framework Refactoring Script
Applies proper console-capture + report-enhancer pattern to all spec files
"""

import os
import re
from pathlib import Path

SPEC_DIR = Path(__file__).parent.parent / 'tests' / 'specFiles' / 'ga'

stats = {
    'total': 0,
    'fixed': 0,
    'skipped': 0,
    'errors': 0,
}

def get_import_depth(file_path):
    """Determine relative import path depth"""
    parts = str(file_path).split(os.sep)
    ga_index = parts.index('ga') if 'ga' in parts else -1
    if ga_index >= 0 and len(parts) - ga_index > 3:  # ga/component/file.spec.ts = 3 parts
        return '../../'
    return '../../../'

def fix_spec_file(file_path):
    """Add ConsoleCapture + report-enhancer to a spec file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        original = content

        # 1. Add report-enhancer import if missing
        if 'attachConsoleCapture' not in content:
            depth = get_import_depth(file_path)
            import_line = f"import {{ attachConsoleCapture, annotateEnvironment }} from '{depth}utils/infra/report-enhancer';"

            # Find last import line
            import_matches = list(re.finditer(r"^import .* from ['\"].+['\"];?$", content, re.MULTILINE))
            if import_matches:
                last_import = import_matches[-1]
                insert_pos = last_import.end()
                content = content[:insert_pos] + '\n' + import_line + content[insert_pos:]

        # 2. Add capture variable if missing
        if 'let capture: ConsoleCapture;' not in content:
            # Find insertion point (after imports, before first const or test)
            pattern = r'\nconst [A-Z_]+ = |^test\.'
            match = re.search(pattern, content, re.MULTILINE)
            if match:
                insert_pos = match.start()
                content = content[:insert_pos] + '\nlet capture: ConsoleCapture;\n' + content[insert_pos:]

        # 3. Update beforeEach to initialize capture
        before_each_pattern = r'test\.beforeEach\s*\(\s*async\s*\(\s*{\s*page\s*}\s*\)\s*=>\s*{([^}]*(?:{[^}]*}[^}]*)*)'
        match = re.search(before_each_pattern, content, re.DOTALL)
        if match and 'capture = new ConsoleCapture' not in match.group(1):
            body = match.group(1)
            # Add capture init before closing
            capture_init = '\n  capture = new ConsoleCapture(page);\n  capture.start();'
            body_fixed = body.rstrip() + capture_init
            content = content[:match.start(1)] + body_fixed + content[match.end(1):]

        # 4. Add afterEach if missing
        if 'test.afterEach' not in content:
            # Find where to add (after beforeEach)
            before_each_end = content.find('});', content.find('test.beforeEach'))
            if before_each_end > 0:
                after_each_block = '''

test.afterEach(async ({ page }, testInfo) => {
  const errors = capture.getErrors();
  const warnings = capture.getWarnings();
  if (errors.length > 0 || warnings.length > 0) {
    await attachConsoleCapture(page, testInfo, errors, warnings);
  }
  await annotateEnvironment(page, testInfo);
});'''
                content = content[:before_each_end + 3] + after_each_block + content[before_each_end + 3:]

        if content != original:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            stats['fixed'] += 1
            print(f"✓ Fixed: {file_path.relative_to(SPEC_DIR.parent)}")
        else:
            stats['skipped'] += 1
    except Exception as e:
        stats['errors'] += 1
        print(f"✗ Error: {file_path}: {e}")

def main():
    print("🔧 Framework Refactoring Script\n")
    print(f"Scanning: {SPEC_DIR}\n")

    # Find all .author.spec.ts files
    author_specs = list(SPEC_DIR.rglob('*.author.spec.ts'))

    print(f"Found {len(author_specs)} author specs\n")

    for spec in author_specs:
        stats['total'] += 1
        fix_spec_file(spec)

    # Print summary
    print(f"\n{'=' * 60}")
    print(f"📊 Summary:")
    print(f"   Total scanned:  {stats['total']}")
    print(f"   Fixed:          {stats['fixed']}")
    print(f"   Skipped:        {stats['skipped']}")
    print(f"   Errors:         {stats['errors']}")
    print(f"{'=' * 60}")

if __name__ == '__main__':
    main()
